File: 02.Queue/01.3.TreeNext.java
/*
// 测试平台链接：https://leetcode-cn.com/problems/populating-next-right-pointers-in-each-node-ii/submissions/
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        Node Q = null;
        if (root != null) {
            Q = root;
        }

        while (Q != null) {
            // 下一层前驱结点
            Node nextLevelPreNode = null;
            // 下一层的头结点
            Node nextLevelHead = null;
            // 顺序遍历当前层的每个结点
            Node curLevelNode = Q;
            while (curLevelNode != null) {
                // 如果得到一个下一层的结点
                if (curLevelNode.left != null) {
                    // 让下一层的前驱结点指向得到的下一层结点
                    if (nextLevelPreNode != null) {
                        nextLevelPreNode.next = curLevelNode.left;
                    }
                    nextLevelPreNode = curLevelNode.left;
                    // 设置下一层的头
                    if (nextLevelHead == null) {
                        nextLevelHead = curLevelNode.left;
                    }
                }
                // 如果得到一个下一层的结点
                if (curLevelNode.right != null) {
                    // 让下一层的前驱结点指向得到的下一层结点
                    if (nextLevelPreNode != null) {
                        nextLevelPreNode.next = curLevelNode.right;
                    }
                    nextLevelPreNode = curLevelNode.right;
                    // 设置下一层的头
                    if (nextLevelHead == null) {
                        nextLevelHead = curLevelNode.right;
                    }
                }
                curLevelNode = curLevelNode.next;
            }
            Q = nextLevelHead;
        }
        return root;
    }
}

File: 02.Queue/01.3.TreeNext.py
"""
测试链接 https://leetcode-cn.com/problems/populating-next-right-pointers-in-each-node-ii/submissions/
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution(object):
    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        Q = None
        if root:
            Q = root

        while Q:
            # 下一层结点的前一个结点
            nextLevelPreNode = None
            # 下一层结点的头结点
            nextLevelHead = None

            # Q是当前层的head
            p = Q
            # 按照顺序依次遍历当前层
            while p:
                # 拿到下一层的结点
                if p.left:
                    # 设置下一层的头
                    if not nextLevelHead:
                        nextLevelHead = p.left
                    # 前下一层的前驱结点的next指向下一层的结点
                    if nextLevelPreNode:
                        nextLevelPreNode.next = p.left
                    nextLevelPreNode = p.left
                if p.right:
                    if not nextLevelHead:
                        nextLevelHead = p.right
                    if nextLevelPreNode:
                        nextLevelPreNode.next = p.right
                    nextLevelPreNode = p.right
                p = p.next
            # 指向下一层的头
            Q = nextLevelHead

        return root
